java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java

```diff
@@ -0,0 +1,141 @@
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.MethodInvocationMatcher;
+import org.sonar.java.checks.methods.TypeCriteria;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.ForEachStatement;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import javax.annotation.CheckForNull;
+
+import java.util.List;
+
+@Rule(
+  key = "S2864",
+  name = "\"entrySet()\" should be iterated when both the key and value are needed",
+  tags = {"performance"},
+  priority = Priority.MAJOR)
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.CPU_EFFICIENCY)
+@SqaleConstantRemediation("5min")
+public class KeySetInsteadOfEntrySetCheck extends SubscriptionBaseVisitor {
+
+  private static final MethodInvocationMatcher MAP_GET_METHOD = MethodInvocationMatcher.create()
+    .typeDefinition(TypeCriteria.subtypeOf("java.util.Map"))
+    .name("get")
+    .addParameter("java.lang.Object");
+
+  private static final MethodInvocationMatcher MAP_KEYSET_METHOD = MethodInvocationMatcher.create()
+    .typeDefinition(TypeCriteria.subtypeOf("java.util.Map"))
+    .name("keySet");
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.FOR_EACH_STATEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (hasSemantic()) {
+      ForEachStatement forEachTree = (ForEachStatement) tree;
+      ExpressionTree expressionTree = forEachTree.expression();
+      if (expressionTree.is(Tree.Kind.METHOD_INVOCATION)) {
+        MethodInvocationTree methodTree = (MethodInvocationTree) expressionTree;
+        Symbol mapSymbol = getMapSymbol(methodTree);
+        if (mapSymbol != null && MAP_KEYSET_METHOD.matches(methodTree)) {
+          new GetUsageVisitor().isCallingGetWithSymbol(forEachTree, forEachTree.variable().symbol(), mapSymbol);
+        }
+      }
+    }
+  }
+
+  @CheckForNull
+  private Symbol getMapSymbol(MethodInvocationTree tree) {
+    ExpressionTree expressionTree = tree.methodSelect();
+    // direct invocation: symbol is implicitly this
+    if (expressionTree.is(Tree.Kind.IDENTIFIER)) {
+      return tree.symbol().owner();
+    }
+    expressionTree = ((MemberSelectExpressionTree) expressionTree).expression();
+    if (expressionTree.is(Tree.Kind.IDENTIFIER)) {
+      return ((IdentifierTree) expressionTree).symbol();
+    } else {
+      return isFieldAccessedUsingSuperOrThis(expressionTree);
```

Comment:
To me, based on its name, the return type of the method should be a Boolean. It should be renamed.